### PR TITLE
feat(EC2): New EC2 AWS check (#852)

### LINKED ISSUE #852: Enhance secrets group: Find secrets in EC2 Auto Scaling Launch Template
I'm having issues with "[extra775] Find secrets in EC2 Auto Scaling Launch Configuration".
I checked the script of extra775, and I think it uses the secretsDetector to check the UserData inside the launch-configuration, but I also feel that there should be a rule to check the UserData inside the launch-template. However, I feel a rule to check UserData in launch-template is also necessary.
(I feel that creating a new one with different rules would be a good idea.)

Reference:
https://docs.aws.amazon.com/autoscaling/ec2/userguide/LaunchTemplates.html

> We recommend that you use launch templates to ensure that you're accessing the latest features and improvements. Not all Amazon EC2 Auto Scaling features are available when you use launch configurations. Not all Amazon EC2 Auto Scaling features are available when you use launch configurations.

### PR DESCRIPTION
**Pull request by rieck-srlabs**
Adds check to find exposed secrets in EC2 Launch Templates. Currently, Prowler does not scan EC2 Launch Templates.

### Context

The existing check `autoscaling_find_secrets_ec2_launch_configuration` only considers EC2 AutoScaling launch configurations, [which are deprecated and can no longer be created](https://docs.aws.amazon.com/autoscaling/ec2/userguide/launch-configurations.html). **The replacement feature - Launch Templates - is currently not checked by Prowler**. However, just like launch configurations, launch templates can also contain `UserData` that might include secrets.

Issue #852 raised this issue in 2021 but failed to result in relevant changes to Prowler. The issue was closed due to inactivity.

### Description

- Updates the EC2 service to include launch templates, including all versions.
- Adds check to search for secrets in EC2 launch templates
- Adds relevant tests for EC2 service and added check
- Adds new check to relevant compliance benchmarks, where `autoscaling_find_secrets_ec2_launch_configuration` was included already.
- Updates the `pentesting.md` tutorial document.

Fixes #852 

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jfagoagas** [comment]
@rieck-srlabs We will try to review this tomorrow or next Monday! Thanks for this contribution!!

**Comment by rieck-srlabs** [comment]
> @rieck-srlabs We will try to review this tomorrow or next Monday! Thanks for this contribution!!

Sounds good. One thing to keep an eye out for: I am not assigning a resource ARN to findings in the newly added check. Launch templates do not have a resource ARN and I was not sure what to use in its place here. Open to suggestions!

**Comment by jfagoagas** [comment]
> Sounds good. One thing to keep an eye out for: I am not assigning a resource ARN to findings in the newly added check. Launch templates do not have a resource ARN and I was not sure what to use in its place here. Open to suggestions!

As per [AWS Documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/permissions-for-launch-templates.html) a Launch Template has the following ARN: `arn:aws:ec2:region:account-id:launch-template/lt-09477bcd97b0d310e`
Screenshot 2024-05-24 at 09 39 39

Could you please include that in the PR? Thanks!

**Comment by rieck-srlabs** [comment]
Good catch, thanks! With the latest commit, the ARN is included now.

File: prowler/providers/aws/services/ec2/ec2_find_secrets_launch_template/ec2_find_secrets_launch_template.metadata.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  "CheckID": "ec2_find_secrets_launch_template",
+  "CheckTitle": "Find secrets in EC2 Launch Template",
+  "CheckType": [
+    "IAM"
+  ],
+  "ServiceName": "ec2",
+  "SubServiceName": "",
+  "ResourceIdTemplate": "arn:partition:service:region:account-id:",
+  "Severity": "critical",
+  "ResourceType": "Other",
```
Comment (MrCloudSec) [inline prowler/providers/aws/services/ec2/ec2_find_secrets_launch_template/ec2_find_secrets_launch_template.metadata.json]:
```suggestion
  "ResourceType": "AwsEc2LaunchTemplate",
```

Comment (MrCloudSec) [reply prowler/providers/aws/services/ec2/ec2_find_secrets_launch_template/ec2_find_secrets_launch_template.metadata.json]: According to https://docs.aws.amazon.com/securityhub/latest/userguide/asff-resourcedetails-awsec2.html#asff-resourcedetails-awsec2launchtemplate

File: prowler/providers/aws/services/ec2/ec2_service.py [hunk 2]
```diff
@@ -470,6 +475,52 @@ def __get_resources_for_regions__(self, regional_client):
...
+                "describe_launch_templates"
+            )
+
+            for page in describe_launch_templates_paginator.paginate():
+                for template in page["LaunchTemplates"]:
+                    self.launch_templates.append(
+                        LaunchTemplate(
+                            name=template["LaunchTemplateName"],
+                            id=template["LaunchTemplateId"],
+                            arn=f"arn:aws:ec2:{regional_client.region}:{self.audited_account}:launch-template/{template['LaunchTemplateId']}",
```
Comment (MrCloudSec) [inline prowler/providers/aws/services/ec2/ec2_service.py]:
```suggestion
                for template in page["LaunchTemplates"]:
                template_arn = f"arn:aws:ec2:{regional_client.region}:{self.audited_account}:launch-template/{template['LaunchTemplateId']}"
                if not self.audit_resources or (
                    is_resource_filtered(template_arn, self.audit_resources)
                ):
                    self.launch_templates.append(
                        LaunchTemplate(
                            name=template["LaunchTemplateName"],
                            id=template["LaunchTemplateId"],
                            arn=template_arn,
```

Comment (MrCloudSec) [reply prowler/providers/aws/services/ec2/ec2_service.py]: We have to add the resource filter function

**Review by MrCloudSec — CHANGES_REQUESTED** [review]
Hi @rieck-srlabs, thanks for this new check! Please, review my comments. 
Could you also change the check name to `ec2_launch_template_no_secrets`? This follows the actual check pattern. Thanks!

**Comment by rieck-srlabs** [comment]
Thanks for the suggestions! I've addressed your comments with my last three commits!

File: prowler/providers/aws/services/ec2/ec2_find_secrets_launch_template/ec2_find_secrets_launch_template.metadata.json (hunk 1 above)
Comment (rieck-srlabs) [reply prowler/providers/aws/services/ec2/ec2_find_secrets_launch_template/ec2_find_secrets_launch_template.metadata.json]: Added in https://github.com/prowler-cloud/prowler/pull/4076/commits/a95f0df83720246fbf72b3f6a13143ca28079463

File: prowler/providers/aws/services/ec2/ec2_service.py (hunk 2 above)
Comment (rieck-srlabs) [reply prowler/providers/aws/services/ec2/ec2_service.py]: Addressed with https://github.com/prowler-cloud/prowler/pull/4076/commits/5ce3d0499b167cb0409905539e972a0703820b2d

File: prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
+{
+  "Provider": "aws",
+  "CheckID": "ec2_launch_template_no_secrets",
+  "CheckTitle": "Find secrets in EC2 Launch Template",
+  "CheckType": [
+    "IAM"
```
Comment (MrCloudSec) [inline prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json]: Why IAM?

File: prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        for template in ec2_client.launch_templates:
+            report = Check_Report_AWS(self.metadata())
+            report.region = template.region
+            report.resource_id = template.id
+            report.resource_arn = template.arn
+
+            versions_with_secrets = []
+
+            for version in template.versions:
+                print(version.template_data)
```
Comment (MrCloudSec) [inline prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.py]:
```suggestion
```

**Review by MrCloudSec — CHANGES_REQUESTED** [review]
Thanks for the changes! I have added a couple of last comments.

File: prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json (hunk 3 above)
Comment (rieck-srlabs) [reply prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json]: [This is the CheckType also used in the existing `ec2_instance_secrets_user_data.metadata` check.](https://github.com/prowler-cloud/prowler/blob/ca554ad3ffee245c1ab87852c83a9987a8ea2f05/prowler/providers/aws/services/ec2/ec2_instance_secrets_user_data/ec2_instance_secrets_user_data.metadata.json#L6)  Open for suggestions here!

File: prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.py (hunk 4 above)
Comment (rieck-srlabs) [reply prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.py]: Thanks, missed this while cleaning up the code.

File: prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json (hunk 3 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json]: I would prefer leaving empty, thanks!

Comment (rieck-srlabs) [reply prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json]: Done with the last commit.

File: prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
+{
+  "Provider": "aws",
+  "CheckID": "ec2_launch_template_no_secrets",
+  "CheckTitle": "Find secrets in EC2 Launch Template",
+  "CheckType": [],
+  "ServiceName": "ec2",
+  "SubServiceName": "",
+  "ResourceIdTemplate": "arn:partition:service:region:account-id:",
```
Comment (MrCloudSec) [inline prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json]:
```suggestion
  "ResourceIdTemplate": "arn:partition:ec2:region:account-id:launch-template/template-id",
```

File: prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  "Severity": "critical",
+  "ResourceType": "AwsEc2LaunchTemplate",
+  "Description": "Find secrets in EC2 Launch Template",
+  "Risk": "The use of a hard-coded password increases the possibility of password guessing.  If hard-coded passwords are used, it is possible that malicious users gain access through the account in question.",
+  "RelatedUrl": "",
```
Comment (MrCloudSec) [inline prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json]:
```suggestion
  "RelatedUrl": "https://docs.aws.amazon.com/AWSEC2/latest/WindowsGuide/ec2-launch-templates.html",
```

Comment (rieck-srlabs) [reply prowler/providers/aws/services/ec2/ec2_launch_template_no_secrets/ec2_launch_template_no_secrets.metadata.json]: I saw that this URL explicitly references Windows. I'm instead resolving this with the more generic `https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-launch-templates.html` URL in the next commit.

**Review by MrCloudSec — APPROVED** [review]
Thanks for the PR and the changes!

**Pull request merged**